### PR TITLE
sasl: binary messages

### PR DESCRIPTION
**Pull request by monnerat**
Capabilities of sasl module are extended to exchange messages in binary as an alternative to base64.

If http authentication flags have been set, those are used as sasl default preferred mechanisms.

### CODE REVIEW

File: lib/curl_sasl.c [hunk 1]
```diff
@@ -58,7 +58,7 @@
 static const struct {
   const char   *name;  /* Name */
   size_t        len;   /* Name length */
-  unsigned int  bit;   /* Flag bit */
+  unsigned long bit;   /* Flag bit */
```
Comment (bagder) [inline lib/curl_sasl.c:61]: Why this change? We cannot assume that a long is larger than 32 bit anyway... and if you want to store 32 bit, why use 64 bit on some platforms?

File: lib/curl_sasl.h [hunk 2]
```diff
@@ -90,27 +92,31 @@ typedef enum {
...
 struct SASL {
   const struct SASLproto *params; /* Protocol dependent parameters */
   saslstate state;         /* Current machine state */
-  unsigned int authmechs;  /* Accepted authentication mechanisms */
-  unsigned int prefmech;   /* Preferred authentication mechanism */
-  unsigned int authused;   /* Auth mechanism used for the connection */
+  const char *curmech;     /* Current mechanism id. */
+  unsigned long authmechs; /* Accepted authentication mechanisms */
+  unsigned long prefmech;  /* Preferred authentication mechanism */
+  unsigned long authused;  /* Auth mechanism used for the connection */
```
Comment (bagder) [inline lib/curl_sasl.h:119]: Same here basically.

File: lib/curl_sasl.c (hunk 1 above)
Comment (monnerat) [reply lib/curl_sasl.c:61]: HTTP authentication flags are already defined as unsigned long: although not yet related, the idea behind this change is to ease a possible future unification of sasl and http authentication flags.
This is the only reason. But perhaps the change would have to be done the other way on http side, which is not the target of my current work.
I can drop it for now, if you prefer.

File: lib/curl_sasl.h (hunk 2 above)
Comment (monnerat) [reply lib/curl_sasl.h]: Last rebase with mechanism flags stored in unsigned ints.

Comment (monnerat) [reply lib/curl_sasl.h]: Outdated: these flags are now stored in unsigned shorts.

**Comment by monnerat** [comment]
IMHO, this is ready for review/pull.

**Comment by monnerat** [comment]
@bagder : Please can you look at this while the feature window is open? Having it unpushed blocks my work on openldap. Thanks.

**Comment by bagder** [comment]
@monnerat for the future, please request a review of the PR. It's an excellent way to signal that you're ready/done (and a good way for me to spot/filter such PRs).

File: lib/pop3.c [hunk 3]
```diff
@@ -171,12 +172,15 @@ const struct Curl_handler Curl_handler_pop3s = {
...
   "pop",                      /* The service name */
-  '*',                        /* Code received when continuation is expected */
-  '+',                        /* Code to receive upon authentication success */
-  255 - 8,                    /* Maximum initial response length (no max) */
   pop3_perform_auth,          /* Send authentication command */
   pop3_continue_auth,         /* Send authentication continuation */
-  pop3_get_message            /* Get SASL response message */
+  pop3_cancel_auth,           /* Send authentication cancellation */
+  pop3_get_message,           /* Get SASL response message */
+  255 - 8,                    /* Maximum initial response length (no max) */
```
Comment (bagder) [inline lib/pop3.c:179]: This comment is a bit strange. What does `(no max)` mean? Is this a copy-paste mistake? Maybe the comment should explain where 255 and 8 come from?

File: lib/imap.c [hunk 4]
```diff
@@ -180,12 +181,15 @@ const struct Curl_handler Curl_handler_imaps = {
...
   "imap",                     /* The service name */
-  '+',                        /* Code received when continuation is expected */
-  IMAP_RESP_OK,               /* Code to receive upon authentication success */
-  0,                          /* Maximum initial response length (no max) */
   imap_perform_authenticate,  /* Send authentication command */
   imap_continue_authenticate, /* Send authentication continuation */
-  imap_get_message            /* Get SASL response message */
+  imap_cancel_authenticate,   /* Send authentication cancellation */
+  imap_get_message,           /* Get SASL response message */
+  0,                          /* Maximum initial response length (no max) */
```
Comment (bagder) [inline lib/imap.c:188]: Does the comment mean that zero means there's no max? Then maybe it should spell it out better?

File: lib/smtp.c [hunk 5]
```diff
@@ -176,12 +177,15 @@ const struct Curl_handler Curl_handler_smtps = {
...
   "smtp",                     /* The service name */
-  334,                        /* Code received when continuation is expected */
-  235,                        /* Code to receive upon authentication success */
-  512 - 8,                    /* Maximum initial response length (no max) */
   smtp_perform_auth,          /* Send authentication command */
   smtp_continue_auth,         /* Send authentication continuation */
-  smtp_get_message            /* Get SASL response message */
+  smtp_cancel_auth,           /* Cancel authentication */
+  smtp_get_message,           /* Get SASL response message */
+  512 - 8,                    /* Maximum initial response length */
```
Comment (bagder) [inline lib/smtp.c:184]: Maybe explain where 512 and 8 come from?

**Comment by monnerat** [comment]
> for the future, please request a review of the PR. It's an excellent way to signal that you're ready/done (and a good way for me to spot/filter such PRs).

Sorry for using a comment for that, but the "request review" icon is sometimes missing. Github bug ?
Anyway I'll use it when present.

File: lib/pop3.c (hunk 3 above)
Comment (monnerat) [reply lib/pop3.c]: Yes, cutn'paste too much!
Done, + a comment in curl_mime.h.

File: lib/imap.c (hunk 4 above)
Comment (monnerat) [reply lib/imap.c]: Done.

File: lib/smtp.c (hunk 5 above)
Comment (monnerat) [reply lib/smtp.c]: Done.

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Comment by monnerat** [comment]
Thanks to you for pulling.
